From da52bb78c689241256399c29a03ae478df057936 Mon Sep 17 00:00:00 2001
From: Oliver Beard <olib141@outlook.com>
Date: Fri, 16 Jan 2026 20:58:00 +0000
Subject: [PATCH 1/3] FROMLIST: lookandfeel/components: Split greeter logic
 from wallpaper fader; use underlay rather than shader

Splitting greeter logic out into a separate component allows usage in Plasma Login Manager, in which the wallpaper lives in a separate binary to the greeter.

Using an underlay rather than a the more complex shader makes things more maintainable, and brings us in-line with other aspects of the shell that want to adjust the wallpaper to create contrast, such as the Overview effect, which uses a blur with radius 64px and a background-coloured underlay to ensure good contrast.

Link: https://invent.kde.org/plasma/plasma-workspace/-/merge_requests/6197
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 lookandfeel/CMakeLists.txt                 |   3 +-
 lookandfeel/components/GreeterFader.qml    |  67 ++++++++++++++
 lookandfeel/components/WallpaperFader.frag |  22 -----
 lookandfeel/components/WallpaperFader.qml  | 100 +++------------------
 4 files changed, 79 insertions(+), 113 deletions(-)
 create mode 100644 lookandfeel/components/GreeterFader.qml
 delete mode 100644 lookandfeel/components/WallpaperFader.frag

diff --git a/lookandfeel/CMakeLists.txt b/lookandfeel/CMakeLists.txt
index 1b76309531..f470e406b9 100644
--- a/lookandfeel/CMakeLists.txt
+++ b/lookandfeel/CMakeLists.txt
@@ -16,6 +16,7 @@ ecm_target_qml_sources(components SOURCES
     components/VirtualKeyboard_wayland.qml
     components/VirtualKeyboardLoader.qml
     components/WallpaperFader.qml
+    components/GreeterFader.qml
 )
 target_sources(components PRIVATE components/componentsplugin.cpp)
 target_link_libraries(components PRIVATE Plasma::Plasma)
@@ -28,8 +29,6 @@ qt6_add_shaders(components LNFSHADERS
     PREFIX "/qt/qml/org/kde/breeze/components/shaders/"
     FILES
         components/UserDelegate.frag
-        components/WallpaperFader.frag
     OUTPUTS
         UserDelegate.frag.qsb
-        WallpaperFader.frag.qsb
 )
diff --git a/lookandfeel/components/GreeterFader.qml b/lookandfeel/components/GreeterFader.qml
new file mode 100644
index 0000000000..458cc412db
--- /dev/null
+++ b/lookandfeel/components/GreeterFader.qml
@@ -0,0 +1,67 @@
+/*
+    SPDX-FileCopyrightText: 2014 Aleix Pol Gonzalez <aleixpol@blue-systems.com>
+
+    SPDX-License-Identifier: GPL-2.0-or-later
+*/
+
+import QtQuick
+import Qt5Compat.GraphicalEffects
+
+import org.kde.kirigami as Kirigami
+
+Item {
+    id: greeterFader
+
+    required property Item mainStack
+    required property Item footer
+    required property Item clock
+
+    required property bool alwaysShowClock
+
+    state: "off"
+
+    states: [
+        State {
+            name: "on"
+            PropertyChanges {
+                mainStack.opacity: 1
+                footer.opacity: 1
+                clock.opacity: 1
+                // Intentional: shadow is not needed due to
+                // WallpaperFader providing enhanced contrast
+                clock.shadow.opacity: 0
+            }
+        },
+        State {
+            name: "off"
+            PropertyChanges {
+                mainStack.opacity: 0
+                footer.opacity: 0
+                clock.opacity: greeterFader.alwaysShowClock ? 1 : 0
+                clock.shadow.opacity: greeterFader.alwaysShowClock ? 1 : 0
+            }
+        }
+    ]
+
+    transitions: [
+        Transition {
+            // NOTE: can't use animators as they don't play well with parallelanimations
+            NumberAnimation {
+                targets: [greeterFader.mainStack, greeterFader.footer, greeterFader.clock]
+                property: "opacity"
+                duration: Kirigami.Units.veryLongDuration
+                easing.type: Easing.InOutQuad
+            }
+        },
+        Transition {
+            NumberAnimation {
+                targets: [greeterFader.clock.shadow]
+                property: "opacity"
+                // Intentional: matches WallpaperFader animation because we only
+                // want shadow for enhanced contrast when wallpaper unfaded
+                duration: Kirigami.Units.veryLongDuration * 2
+                easing.type: Easing.InOutQuad
+            }
+        }
+    ]
+}
diff --git a/lookandfeel/components/WallpaperFader.frag b/lookandfeel/components/WallpaperFader.frag
deleted file mode 100644
index 736860ad5a..0000000000
--- a/lookandfeel/components/WallpaperFader.frag
+++ /dev/null
@@ -1,22 +0,0 @@
-/*
-    SPDX-FileCopyrightText: 2014 Aleix Pol Gonzalez <aleixpol@blue-systems.com>
-    SPDX-FileCopyrightText: 2023 David Redondo <kde@david-redondo.de>
-
-    SPDX-License-Identifier: GPL-2.0-or-later
-*/
-#version 440
-
-layout(location = 0) in vec2 qt_TexCoord0;
-layout(std140, binding = 0) uniform buf {
-    mat4 qt_Matrix;
-    float qt_Opacity;
-    mat4 colorMatrix;
-};
-layout(binding = 1) uniform mediump sampler2D source;
-layout(location = 0) out vec4 fragColor;
-
-void main(void)
-{
-    mediump vec4 tex = texture(source, qt_TexCoord0);
-    fragColor = tex * colorMatrix * qt_Opacity;
-}
diff --git a/lookandfeel/components/WallpaperFader.qml b/lookandfeel/components/WallpaperFader.qml
index 2d9cc2d860..0ed582c399 100644
--- a/lookandfeel/components/WallpaperFader.qml
+++ b/lookandfeel/components/WallpaperFader.qml
@@ -11,16 +11,13 @@ import org.kde.kirigami as Kirigami
 
 Item {
     id: wallpaperFader
-    property Item clock
-    property Item mainStack
-    property Item footer
-    property alias source: wallpaperBlur.source
-    property real factor: 0
-    readonly property bool lightColorScheme: Math.max(Kirigami.Theme.backgroundColor.r, Kirigami.Theme.backgroundColor.g, Kirigami.Theme.backgroundColor.b) > 0.5
 
-    property bool alwaysShowClock: false
+    required property real factor
+    property alias source: wallpaperBlur.source
 
-    state: "on"
+    /* These values match those used in the Plasma overview effect */
+    readonly property int blurRadius: 64
+    readonly property real underlayOpacity: 0.7
 
     Behavior on factor {
         NumberAnimation {
@@ -30,92 +27,17 @@ Item {
             easing.type: Easing.InOutQuad
         }
     }
+
     FastBlur {
         id: wallpaperBlur
         anchors.fill: parent
-        radius: 50 * wallpaperFader.factor
+        radius: wallpaperFader.blurRadius * wallpaperFader.factor
     }
-    ShaderEffect {
-        id: wallpaperShader
-        anchors.fill: parent
-        supportsAtlasTextures: true
-        property var source: ShaderEffectSource {
-            sourceItem: wallpaperBlur
-            live: true
-            hideSource: true
-            textureMirroring: ShaderEffectSource.NoMirroring
-        }
-
-        readonly property real contrast: 0.8 * wallpaperFader.factor + (1 - wallpaperFader.factor)
-        readonly property real saturation: 1.5 * wallpaperFader.factor + (1 - wallpaperFader.factor)
-        readonly property real intensity: (wallpaperFader.lightColorScheme ? 1.6 : 0.7) * wallpaperFader.factor + (1 - wallpaperFader.factor)
 
-        readonly property real transl: (1.0 - contrast) / 2.0;
-        readonly property real rval: (1.0 - saturation) * 0.2126;
-        readonly property real gval: (1.0 - saturation) * 0.7152;
-        readonly property real bval: (1.0 - saturation) * 0.0722;
-
-        property var colorMatrix: Qt.matrix4x4(
-            contrast, 0,        0,        0.0,
-            0,        contrast, 0,        0.0,
-            0,        0,        contrast, 0.0,
-            transl,   transl,   transl,   1.0).times(Qt.matrix4x4(
-                rval + saturation, rval,     rval,     0.0,
-                gval,     gval + saturation, gval,     0.0,
-                bval,     bval,     bval + saturation, 0.0,
-                0,        0,        0,        1.0)).times(Qt.matrix4x4(
-                    intensity, 0,         0,         0,
-                    0,         intensity, 0,         0,
-                    0,         0,         intensity, 0,
-                    0,         0,         0,         1
-                ));
+    Rectangle {
+        anchors.fill: parent
 
-        fragmentShader: "qrc:/qt/qml/org/kde/breeze/components/shaders/WallpaperFader.frag.qsb"
+        opacity: factor * wallpaperFader.underlayOpacity
+        color: Kirigami.Theme.backgroundColor
     }
-
-    states: [
-        State {
-            name: "on"
-            PropertyChanges {
-                mainStack.opacity: 1
-                footer.opacity: 1
-                wallpaperFader.factor: 1
-                clock.shadow.opacity: 0
-                clock.opacity: 1
-            }
-        },
-        State {
-            name: "off"
-            PropertyChanges {
-                mainStack.opacity: 0
-                footer.opacity: 0
-                wallpaperFader.factor: 0
-                clock.shadow.opacity: wallpaperFader.alwaysShowClock ? 1 : 0
-                clock.opacity: wallpaperFader.alwaysShowClock ? 1 : 0
-            }
-        }
-    ]
-    transitions: [
-        Transition {
-            from: "off"
-            to: "on"
-            //Note: can't use animators as they don't play well with parallelanimations
-            NumberAnimation {
-                targets: [wallpaperFader.mainStack, wallpaperFader.footer, wallpaperFader.clock]
-                property: "opacity"
-                duration: Kirigami.Units.veryLongDuration
-                easing.type: Easing.InOutQuad
-            }
-        },
-        Transition {
-            from: "on"
-            to: "off"
-            NumberAnimation {
-                targets: [wallpaperFader.mainStack, wallpaperFader.footer, wallpaperFader.clock]
-                property: "opacity"
-                duration: Kirigami.Units.veryLongDuration
-                easing.type: Easing.InOutQuad
-            }
-        }
-    ]
 }
-- 
2.52.0

